libsolidity/codegen/ir/IRGeneratorForStatements.cpp

```diff
@@ -1361,26 +1394,143 @@ void IRGeneratorForStatements::appendAndOrOperatorCode(BinaryOperation const& _b
 
 	_binOp.leftExpression().accept(*this);
 
-	string value = m_context.variable(_binOp);
-	m_code << "let " << value << " := " << m_context.variable(_binOp.leftExpression()) << "\n";
+	IRVariable value(_binOp);
+	define(value, _binOp.leftExpression());
 	if (op == Token::Or)
-		m_code << "if iszero(" << value << ") {\n";
+		m_code << "if iszero(" << value.name() << ") {\n";
 	else
-		m_code << "if " << value << " {\n";
+		m_code << "if " << value.name() << " {\n";
 	_binOp.rightExpression().accept(*this);
-	m_code << value << " := " + m_context.variable(_binOp.rightExpression()) << "\n";
+	assign(value, _binOp.rightExpression());
 	m_code << "}\n";
 }
 
-void IRGeneratorForStatements::setLValue(Expression const& _expression, unique_ptr<IRLValue> _lvalue)
+void IRGeneratorForStatements::writeToLValue(IRLValue const& _lvalue, IRVariable const& _value)
+{
+	std::visit(
+		util::GenericVisitor{
+			[&](IRLValue::Storage const& _storage) {
+				std::optional<unsigned> offset;
+
+				if (std::holds_alternative<unsigned>(_storage.offset))
+					offset = std::get<unsigned>(_storage.offset);
+
+				IRVariable prepared(m_context.newYulVariable(), _lvalue.type);
+				define(prepared, _value);
+
+				m_code <<
+					m_utils.updateStorageValueFunction(_lvalue.type, offset) <<
+					"(" <<
+					_storage.slot <<
+					(
+						std::holds_alternative<string>(_storage.offset) ?
+						(", " + std::get<string>(_storage.offset)) :
+						""
+					) <<
+					", " <<
+					prepared.commaSeparatedList() <<
+					")\n";
+			},
+			[&](IRLValue::Memory const& _memory) {
+				if (_lvalue.type.isValueType())
+				{
+					IRVariable prepared(m_context.newYulVariable(), _lvalue.type);
+					define(prepared, _value);
+
+					// TODO: does this actually need cleanup?
+
+					if (_memory.byteArrayElement)
+					{
+						solAssert(_lvalue.type == *TypeProvider::byte(), "");
+						m_code << "mstore8(" + _memory.address + ", byte(0, " + prepared.commaSeparatedList() + "))\n";
+					}
+					else
+						m_code << m_utils.writeToMemoryFunction(_lvalue.type) <<
+							"(" <<
+							_memory.address <<
+							", " <<
+							prepared.commaSeparatedList() <<
+							")\n";
+				}
+				else
+				{
+					solAssert(_lvalue.type.sizeOnStack() == 1, "");
+					solAssert(dynamic_cast<ReferenceType const*>(&_lvalue.type), "");
```

Comment:
Plus asser that `_value` is a memory reference type.

Reply:
I can even assert that ``_value.type() == _lvalue.type``, can't I?

Reply:
Or maybe ``isImplicitlyConvertibleTo``?

Reply:
Not sure about ``string`` and ``bytes``....

Reply:
Well, I'll leave it at asserting it's a memory reference for now.